allennlp_rc/dataset_readers/transformer_squad.py

```diff
@@ -0,0 +1,325 @@
+        if (
+            AutoModel.from_pretrained(
+                transformer_model_name
+            ).embeddings.token_type_embeddings.num_embeddings
+            > 1
+        ):
+            self.non_content_type_id = 1
+        else:
+            self.non_content_type_id = 0
+
+    @overrides
+    def _read(self, file_path: str):
+        # if `file_path` is a URL, redirect to the cache
+        file_path = cached_path(file_path)
+
+        logger.info("Reading file at %s", file_path)
+        open_fn: Callable[[str], IO[Any]] = open
+        if file_path.endswith(".gz"):
+            import gzip
+
+            open_fn = gzip.open
+        elif file_path.endswith(".bz2"):
+            import bz2
+
+            open_fn = bz2.open
+        with open_fn(file_path) as dataset_file:
+            dataset_json = json.load(dataset_file)
+            dataset = dataset_json["data"]
+        logger.info("Reading the dataset")
+        question_count = 0
+        questions_with_more_than_one_instance = 0
+        for article in dataset:
+            for paragraph_json in article["paragraphs"]:
+                context = paragraph_json["context"]
+                for question_answer in paragraph_json["qas"]:
+                    answers = [answer_json["text"] for answer_json in question_answer["answers"]]
+
+                    # This version does not support no-answer questions.
+                    if len(answers) <= 0:
+                        continue
+
+                    # Just like huggingface, we only use the first answer for training.
+                    first_answer_offset = int(question_answer["answers"][0]["answer_start"])
+
+                    instances = self.make_instances(
+                        question_answer.get("id", None),
+                        question_answer["question"],
+                        answers,
+                        context,
+                        first_answer_offset,
+                    )
+                    instances_yielded = 0
+                    for instance in instances:
+                        yield instance
+                        instances_yielded += 1
+                    if instances_yielded > 1:
+                        questions_with_more_than_one_instance += 1
+                    question_count += 1
+                    if self.max_questions is not None and question_count >= self.max_questions:
+                        return
+
+        logger.info(
+            "%d (%.2f%%) questions have more than one instance",
+            questions_with_more_than_one_instance,
+            100 * questions_with_more_than_one_instance / question_count,
+        )
+
+    def make_instances(
+        self, qid: str, question: str, answers: List[str], context: str, first_answer_offset: int
+    ) -> Iterable[Instance]:
+        answer_end = first_answer_offset + len(answers[0])
+
+        # tokenize context by spaces first, and then with the wordpiece tokenizer
```

Comment:
It might be easier to build on top of this? https://github.com/allenai/allennlp/blob/6babeef119ad02ed184e265a57b1df563a0c3376/allennlp/data/tokenizers/pretrained_transformer_tokenizer.py#L181

Reply:
I'm not sure it is. The important thing here is to preserve the character offsets.

Reply:
In fact, the only reason it does it this way is because huggingface does. I think I can get the same score without this extra step.

Reply:
I think it wouldn't be hard to support character offsets in `intra_word_tokenize`, and this entire block of code could simply become `tokenized_context = self.tokenizer.intra_word_tokenize(context)`. But it's fine with me if you think doing it this way would be simpler.

Reply:
If the extra step doesn't contribute anything to performance, I'm definitely in favor of removing it.

Reply:
I'm running without it on `allennlp-server1` to make sure the scores are not affected.

Reply:
When I take this out, it fails with `bert-base` and `bert-large`. The problem is that finding character offsets goes off the rails when the original text reads `"/ˈʃwɔːrtsənˌɛɡər/; German: [ˈaɐ̯nɔlt ˈalɔʏs ˈʃvaɐ̯tsn̩ˌɛɡɐ]"`. The spaces give it a chance to catch back up, while without spaces, once it's lost, it's lost. This is another thing that will hopefully be fixed with the new tokenizers.

On balance, I want to leave the spaces thing here and try this again when we have the new tokenizers.

Reply:
:+1: